new package: gatetools

glennpj opened the pull request:
This PR adds the gatetools package and dependencies. The gatetools
package is a set of command line tools for gate. Since it is primarily a
CLI, although python modules can be loaded, it is named gatetools as
opposed to py-gatetools.

glennpj committed:
new package: gatetools

This PR adds the gatetools package and dependencies. The gatetools
package is a set of command line tools for gate. Since it is primarily a
CLI, although python modules can be loaded, it is named gatetools as
opposed to py-gatetools.

glennpj committed:
Fix quote characterss to avoid test error

glennpj committed:
Found another UTF8 character that was tripping up tests

glennpj committed:
Another UTF-8 character to replace

adamjstewart commented:
I'll try to review this on Tuesday, remind me if I don't get to it.

adamjstewart commented:
P.S. PRs are easier to review if you submit a separate PR for each package.

glennpj commented:
> P.S. PRs are easier to review if you submit a separate PR for each package.

Understood, but managing dependencies in github is much harder. Tests for packages that have dependencies on packages that are not merged yet will fail due to the missing dependency in the repo. 

Would you like me to resubmit with these split up?

adamjstewart commented:
It's not required, just a suggestion. To solve the dependency test issue, you just have to close and reopen the PR after the dependencies are merged. Although this does double the number of tests. But it makes reviewing easier, so it depends on how busy you are and how busy the people reviewing the PRs are. (I'm currently busy but that should end tomorrow).

adamjstewart commented on var/spack/repos/builtin/packages/gatetools/package.py:
```diff
@@ -0,0 +1,37 @@
+    depends_on('py-python-box', type=('build', 'run'))
+    depends_on('gate+rtk', type='run')
+
+    # The readme.md file is not in the distribution, so fake it.
+    @run_before('build')
+    def readme(self):
+        touch('readme.md')
```
This is kind of embarrassing... You might want to tell the developers about this. It's not hard to include this file in the PyPI tarball using this: https://setuptools.readthedocs.io/en/latest/userguide/datafiles.html

adamjstewart commented on var/spack/repos/builtin/packages/gatetools/package.py:
```diff
@@ -0,0 +1,37 @@
+    depends_on('py-tqdm', type=('build', 'run'))
+    depends_on('py-colored', type=('build', 'run'))
+    depends_on('py-itk@5.1.0:', type=('build', 'run'))
+    depends_on('py-uproot3', type=('build', 'run'))
```
```suggestion
    depends_on('py-uproot', type=('build', 'run'))
```
We already have a package for this.

glennpj replied:
It actually needs py-uproot3. I created the py-uproot package thinking it was needed here but then later learned that py-uproot3 was needed instead. I went ahead and sent a PR for py-uproot, #22658, since I already had created the package.

adamjstewart commented on var/spack/repos/builtin/packages/gatetools/package.py:
```diff
@@ -0,0 +1,37 @@
+    depends_on('py-itk@5.1.0:', type=('build', 'run'))
+    depends_on('py-uproot3', type=('build', 'run'))
+    depends_on('py-wget', type=('build', 'run'))
+    depends_on('py-python-box', type=('build', 'run'))
+    depends_on('gate+rtk', type='run')
```
I don't see these dependencies listed anywhere, where did you find them?

glennpj replied:
I got python-box from the setup.py file, but from the master branch of the git repository. It is not listed in the release file. I will remove that and pull out the corresponding package file to submit as a separate PR. Sorry about that, not sure why I was looking at the master branch.

The gate dependency is not listed but is needed, as gatetools needs gate available.

adamjstewart commented on var/spack/repos/builtin/packages/py-awkward0/package.py:
```diff
@@ -0,0 +1,21 @@
+from spack import *
+
+
+class PyAwkward0(PythonPackage):
```
Similar to `py-uproot`, we already have a `py-awkward` package. The `py-awkward0` and `py-awkward1` packages seem to no longer be getting updates on PyPI, so I think it's better to use `py-awkward` instead. Maybe @vvolkl can comment.

glennpj replied:
The py-awkward0 package was released February 8, 2021, which seems pretty recent. I am pretty sure that py-awkward0 is correct.

vvolkl replied:
@glennpj No, the developers are very clear on this one:

>  Fundamental changes were also required to streamline remote reading (HTTP and XRootD), so Uproot 4 was distributed as a separate project in parallel with Uproot 3 (like Awkward 1 and 0). For the latter half of 2020, adventurous users and downstream developers could install uproot as a separate project.
https://raw.githubusercontent.com/scikit-hep/uproot4/main/docs-img/diagrams/uproot-awkward-timeline.png

> On December 1, 2020, however, Awkward 0 and Uproot 3 were deprecated, moved to PyPI packages awkward0 and uproot3, while Awkward 1 and Uproot 4 became unqualified as awkward and uproot.

https://uproot.readthedocs.io/en/latest/uproot3-to-4.html

If you check the commit history of awkward0.x, it  is clear that there are only fixes to help people who need it for legacy code, not new developments. I'd say that if possible, it would be a good idea to avoid adding the old version and see if the downstream package can migrate to the new uproot/awkward. If not, please add a comment that this is a legacy package and that `awkward`/`uproot` should be used for new projects.

glennpj replied:
Here is what I get when using py-uproot4 with gatetools:
```console
$ gt_gate_info 
uproot3 is mandatory to merge root file. Please, do:
pip install uproot3
uproot3 is mandatory to merge root file. Please, do:
pip install uproot3
Gate path:          /opt/packages/gpjohnsn/spack/opt/spack/linux-centos7-broadwell/gcc-8.4.0/gate-9.0-npreog6wphf5eaenfhkuqyp5ezmnqyut/bin/Gate
Version:            Gate version is 9.0 ; Geant4 version is 10.6.3
...
```
So, gatetools needs py-uproot3, and py-uproot3 and py-uproot3-methods need py-awkward0. I am not involved with the OpenGate project. I am just trying to provide spack packages for researchers who use gate and gatetools, so I am not in a position to say whether the gatetools project is or is not looking into updating that dependency. Since the error message above specifically calls out uproot3 it seems that they are aware of the situation. It seems to me that this is a legitimate use of the legacy packages provided by PyPI and Spack should provide them as needed.

vvolkl replied:
Sure, that's what uproot3 is for. I had a quick look and gatetools uses the write functionality in uproot3 that is still not completely ported to uproot, so it is also not trivial to replace it for the developers at the moment.

adamjstewart commented on var/spack/repos/builtin/packages/py-itk/package.py:
```diff
@@ -0,0 +1,106 @@
+from spack import *
+
+
+class PyItk(Package):
```
I hate when libraries only release wheels on PyPI. I know we prefer PyPI when possible, but in this case I would suggest downloading from the [homepage](https://itk.org/download/) or [GitHub](https://github.com/InsightSoftwareConsortium/ITK/releases) instead. That should significantly reduce the complexity of the package.

glennpj replied:
I looked into that but it was kind of messy. It seemed like the wheel from PyPI was the better option, but yes, that is kind of messy as well. The git project that builds the py-itk wheels is something that I looked at, but it only builds wheels so a wheel package would need to be built and installed via spack. That process would also involve downloading itk source. I suppose the build process for that could be hacked on to make it work with spack.

adamjstewart commented on var/spack/repos/builtin/packages/py-pydicom/package.py:
```diff
@@ -0,0 +1,24 @@
+    depends_on('py-setuptools', type='build')
+
+    # py-numpy is not required but is needed for Pixel Data
+    depends_on('py-numpy',        type=('build', 'run'))
```
Should we add a variant for this?

glennpj replied:
I did think about that. My thinking is that numpy is generally useful and pretty common so might as well include it. But I can see it as a variant as well. That is kind of why I separated it out and commented on it. I guess that should be a run dependency though, not a build and run type.

glennpj replied:
After more thought, I think it makes sense to have a variant for this so I will add that.

adamjstewart commented on var/spack/repos/builtin/packages/py-uproot3-methods/package.py:
```diff
@@ -0,0 +1,26 @@
+from spack import *
+
+
+class PyUproot3Methods(PythonPackage):
```
Not needed if you use `py-uproot` instead of `py-uproot3`

glennpj replied:
It is needed, as py-uproot3 is needed.

adamjstewart commented on var/spack/repos/builtin/packages/py-uproot3/package.py:
```diff
@@ -0,0 +1,29 @@
+from spack import *
+
+
+class PyUproot3(PythonPackage):
```
Explained a bit above, but I think you should use `py-uproot` instead of `py-uproot3`. uproot3 and uproot4 seem to no longer be getting updates on PyPI.

glennpj replied:
Yep, per my comments, uproot3 is needed. It does seem to be getting updates, at least as of February 2021.

glennpj committed:
Remove py-python-box dependency and package file

glennpj committed:
Make numpy a variant

- py-setuptools needs to be a run dependendency
  This was masked by py-numpy having py-setuptools as a run dependency.

adamjstewart commented on var/spack/repos/builtin/packages/py-awkward0/package.py:
```diff
@@ -0,0 +1,21 @@
+
+    version('0.15.5', sha256='156e6e338c56d857a7bb53c4fcc8b0b2592a3470eff0d854e6d68777986359ad')
+
+    depends_on('py-setuptools', type='build')
```
Missing build dep on `py-pytest-runner`

adamjstewart commented on var/spack/repos/builtin/packages/py-pydicom/package.py:
```diff
@@ -0,0 +1,24 @@
+
+    version('2.1.2', sha256='65f36820c5fec24b4e7ca45b7dae93e054ed269d55f92681863d39d30459e2fd')
+
+    variant('numpy', default=False, description='Use NumPy for Pixel data')
```
I might suggest the variant to be called `pixel` and the description be "Enable support for Pixel data".

glennpj replied:
That was my initial thought. However, pydicom always works with Pixel data, it is just that if numpy is installed the data can be converted to a numpy array. I felt like naming it 'pixel' gave the wrong impression. I also thought of 'pixel-arrays' but that seemed long.

adamjstewart commented on var/spack/repos/builtin/packages/py-uproot3/package.py:
```diff
@@ -0,0 +1,29 @@
+    version('3.14.4', sha256='4396746ba5ef9071bb0a9da53294e4613a7f4548218940f86496e79d682d20eb')
+
+    depends_on('python@2.7:2.9,3.5:', type=('build', 'run'))
+    depends_on('py-setuptools', type='build')
```
Missing build dep on `py-pytest-runner`

glennpj committed:
Add missing build depency on py-pytest-runner
